app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,491 @@
+        }
+
+        // trim if we fetched new data, and spatialCache is full
+        // trim to 90%, so trim is (probably) not immediately called on next fetch
+        if (spatialCache.size >= maxTiles && tilesToFetch.isNotEmpty())
+            trim((maxTiles * 9) / 10)
+        return result
+    }
+
+    /** Clears the cache */
+    fun clear() = synchronized(this) {
+        spatialCache.clear()
+        wayCache.clear()
+        relationCache.clear()
+        wayGeometryCache.clear()
+        relationGeometryCache.clear()
+        wayIdsByNodeIdCache.clear()
+        relationIdsByElementKeyCache.clear()
+    }
+
+    /** Reduces cache size to the given number of non-empty [tiles], and removes all data
+     *  not contained in the remaining tiles.
+     */
+    fun trim(tiles: Int) = synchronized(this) {
+        spatialCache.trim(tiles)
+        trimNonSpatialCaches()
+    }
+
+    private fun trimNonSpatialCaches() {
+        synchronized(this) {
+            // ways and relations with at least one element in cache should not be removed
+            val (wayIds, relationIds) = getWayAndRelationIdsWithElementsInSpatialCache()
+
+            wayCache.keys.retainAll { it in wayIds }
+            relationCache.keys.retainAll { it in relationIds }
+            wayGeometryCache.keys.retainAll { it in wayIds }
+            relationGeometryCache.keys.retainAll { it in relationIds }
+
+            // now clean up wayIdsByNodeIdCache and relationIdsByElementKeyCache
+            wayIdsByNodeIdCache.keys.retainAll { spatialCache.get(it) != null }
+            relationIdsByElementKeyCache.keys.retainAll {
+                when (it.type) {
+                    ElementType.NODE -> spatialCache.get(it.id) != null
+                    ElementType.WAY -> it.id in wayIds
+                    ElementType.RELATION -> it.id in relationIds
+                }
+            }
+        }
+    }
+
+    private fun getWayAndRelationIdsWithElementsInSpatialCache(): Pair<Set<Long>, Set<Long>> = synchronized(this) {
+        val wayIds = wayCache.values.mapNotNull { way ->
+            if (way != null && way.nodeIds.any { spatialCache.get(it) != null }) way.id
+            else null
+        }.toHashSet()
+
+        fun RelationMember.isCached(): Boolean =
+            type == ElementType.NODE && spatialCache.get(ref) != null
+                || type == ElementType.WAY && ref in wayIds
+
+        val relationIds = relationCache.values.mapNotNull { relation ->
+            if (relation != null && relation.members.any { member ->
+                    member.isCached()
+                        || (member.type == ElementType.RELATION // relation of relations
+                        && (relationCache[member.ref])?.members?.any { it.isCached() } == true)
+                })
+                relation.id
+            else null
+        }.toHashSet()
+        wayIds to relationIds
```

Comment:
This method is actually extremely hard to read. I only roughly understand what it does. Any way it could be improved, even if the code is then longer in terms of lines of code?

Reply:
it's a bit adjusted now, I hope this is better

Reply:
I'll re-read through it, thanks!